make/autoconf/lib-hsdis.m4

```diff
@@ -238,9 +239,10 @@ AC_DEFUN([LIB_SETUP_HSDIS_BINUTILS],
     AC_CHECK_LIB(opcodes, disassembler, [ HSDIS_LIBS="$HSDIS_LIBS -lopcodes" ], [ binutils_system_error="libopcodes not found" ])
     AC_CHECK_LIB(iberty, xmalloc, [ HSDIS_LIBS="$HSDIS_LIBS -liberty" ], [ binutils_system_error="libiberty not found" ])
     AC_CHECK_LIB(z, deflate, [ HSDIS_LIBS="$HSDIS_LIBS -lz" ], [ binutils_system_error="libz not found" ])
+    AC_CHECK_LIB(sframe, frame, [ HSDIS_LIBS="$HSDIS_LIBS -lsframe" ], )
     HSDIS_CFLAGS="-DLIBARCH_$OPENJDK_TARGET_CPU_LEGACY_LIB"
   elif test "x$BINUTILS_DIR" != x; then
-    if test -e $BINUTILS_DIR/bfd/libbfd.a && \
+    if test -e $BINUTILS_DIR/bfd/.libs/libbfd.a && \
```

Comment:
The use of `.libs` here looks a bit scary. I have not looked in detail at the binutils build system, but it sounds like a temporary/internal file. What is the rationale?

Reply:
We never do 'make install' , so we are using the internal place of files.
Make install do not install some of the dependency we have (some h-files we use are not part of binutils installation, e.g. bfdver.h).

So we are stuck between hard place and a rock, we can't use the installed binutils file-tree, so we must keep using the internal build layout.

Reply:
It may be possible to read some elf header about binutils version.
I.e. do make install from our build system, then reading out version from elf section and pass that into our build.

Reply:
I'm still not really happy about this. The old solution without .libs has worked before -- has anything changed in newer versions of binutils? 

Also, we support building binutils in place as a convenience, but it should also be possible to just set a hsdis path, and in that case we cannot presume that the .libs layout is kept.

I suggest you change this to both eat the cake and have it -- first check for the lib in the original location (without .libs), and if it is not found there, check in .libs. This is perhaps not necessary here in LIB_BUILD_BINUTILS, but it definitely is in LIB_SETUP_HSDIS_BINUTILS. It will make the code a few lines longer but more robust.

Reply:
> It may be possible to read some elf header about binutils version. I.e. do make install from our build system, then reading out version from elf section and pass that into our build. As we need to know the signature of init_disassemble_info() which is different between binutils versions.

Trying to understand what you are really saying here. If we accept this patch, it will no longer be possible to use a normally installed binutils, since the BFD_VERSION constant (via bfdver.h) is not available? 

That is not good; the idea of building binutils locally was supposed to be a convenience, not a requirement.

Reply:
Both options --with-binutils= and --with-binutils-src= grab atrifact from temporary internal build layout.
If you do "make install" and point --with-binutils= to that directory the layout is diffrent and you get:
--with-binutils=/home/rehn/binutils-2.39 gives:
configure: error: /home/rehn/binutils-2.39 does not contain a proper binutils installation

We can't use my distro native or a installed binutils.

> I'm still not really happy about this. The old solution without .libs has worked before -- has anything changed in newer versions of binutils?

As the PR says: "libbfd.a is only present in .lib directory in newer binutils builds (older it is in both directories)"
We are depending on their build system artifact layout, this is not a stable interface.
You need todo make install to get the stable layout, then it would be "./lib/libsframe.a".

> 
> Also, we support building binutils in place as a convenience, but it should also be possible to just set a hsdis path, and in that case we cannot presume that the .libs layout is kept.

binutils do not support having source dependency on multiple versions (i asked about this).
That is way the binutils tools are co-hosted (gdb, gas, ...)
Which I assume is on of the reasons it is not possible form headers to check version.
But as I said we today do not support using the installation since we have wrong path for libs.

> 
> I suggest you change this to both eat the cake and have it -- first check for the lib in the original location (without .libs), and if it is not found there, check in .libs. This is perhaps not necessary here in LIB_BUILD_BINUTILS, but it definitely is in LIB_SETUP_HSDIS_BINUTILS. It will make the code a few lines longer but more robust.

As I said it have been under .lib in all version of binutils I checked. This is not a new location.
So it seem like it's much better to check the directory it is always in (so far), no?

> > It may be possible to read some elf header about binutils version. I.e. do make install from our build system, then reading out version from elf section and pass that into our build. As we need to know the signature of init_disassemble_info() which is different between binutils versions.
> 
> Trying to understand what you are really saying here. If we accept this patch, it will no longer be possible to use a normally installed binutils, since the BFD_VERSION constant (via bfdver.h) is not available?
> 
> That is not good; the idea of building binutils locally was supposed to be a convenience, not a requirement.

"--with-binutils" don't use the "normally installed binutils" it assume that the directory is the build directory, which as have a different layout than "make install". So that case do not work today.

Summarize, the layout we are using for both hsdis= and hsdis-src= are the internal temporary layout before you do make install. Therefore we have access to bfdver.h and can determine API version.

This do not handle the case if you built with dependency on zstd.

Reply:
I forgot about --with-binutils=system:
This one is broken:
/home/rehn/source/jdk/open/src/utils/hsdis/binutils/hsdis-binutils.c:60:10: fatal error: libiberty.h: No such file or directory
   60 | #include <libiberty.h>

This include is under "libiberty/libiberty.h".

And yes, if we fix that bfdver.h would be an issue, you are correct.

Maybe we check just test what API version in make system by testing the init_disassemble_info_from_bfd signature.
Thus removing bfdver.h.

For a system build we would have deps on libzstd.so, if it was not compiled without it.